Make the Grid fast again

ericflock opened the pull request:
Make the Grid fast again.

As described in several issues, the Grid render performance is not as good as expected, especially if it's configured with many columns.

This change solves these issues by avoiding loops with multiple calls to the same functions for every column.
This was also introduced in parts by tsuoanttila with https://github.com/vaadin/framework/pull/10579 for Vaadin 8. This change solves these issues in Vaadin 7.7

Additionally removed the column-width calculation in escaltor on paintInsertRows, because the calculation will be anyway scheduled.

Because this is a performance issue, no test classes are provided, instead I provide some benchmark data to show the improvement of this change.

This benchmark was measured with Chrome 67.

[grid-performance.pdf](https://github.com/vaadin/framework/files/2184831/grid-performance.pdf)

As you can see the performance increases significantly.

This is the first change to increase the performance of the Grid in Vaadin 7.7. A second change, which optimizes the automatic column width calculation, will follow in case this pull request will be successfully merged.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/11043)
<!-- Reviewable:end -->

ericflock committed:
Make the Grid fast again.

ericflock committed:
Make the Grid fast again - fix broken column hiding toggles

ericflock committed:
Make the Grid fast again - clean up

ericflock committed:
Make the Grid fast again - java6 compliant - revert unnecessary change

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/framework?pullRequest=11043) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/framework?pullRequest=11043) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/framework?pullRequest=11043) it.</sub>

ericflock committed:
Make the Grid fast again - fix broken editor

ericflock committed:
Make the Grid fast again - fix wrong index

ericflock commented:
@tsuoanttila
Hi, excuse for my questions, but I need little help from you for solving the failing tests?

The most tests failed due to a NoSuchElementException, but I cannot reproduce it on my local machine. The elements with the selectors to find are all there. Could it be a race condition? Furthermore some tests fail, although I changed nothing on the testet component itself, e.g. WebBrowserTimeZoneTest or ComboBoxItemIconTest. Are the tests currently working?

Thanks in advance

tsuoanttila commented:
We had some recent changes in our testing environment and have not had time to fix the issues for Vaadin 7, these problems are being worked on. Once we have them fixed, you'll need to rebase your change on top of latest `7.7`. I will respond here again when the fixes are in place.

tsuoanttila commented:
Validation build has now been fixed, I updated the PR.

ericflock committed:
Make the Grid fast again - fix reattach

ericflock committed:
Make the Grid fast again - revert change to avoid inserting dummy rows

cause it leads to a wrong focused cell, and also no backend data will be
requested initially.

elmot commented:
Should be ported to compatibility if merged

ericflock committed:
Make the Grid fast again - minor refactoring

elmot reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @ericflock)

___
*[client/src/main/java/com/vaadin/client/connectors/GridConnector.java, line 175 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpp9s6WV6y9ObOmY_:-LLEpp9s6WV6y9ObOmYa:b-oisj1v) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/connectors/GridConnector.java#L175)):*
> ```Java
> 
>         @SuppressWarnings("unchecked")
>         public CustomGridColumn(GridColumnState state) {
> ```

API breaking change. Is it possible to keep old constructor parameters and then get the state via `AbstractRendererConnector.getState()`
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6485 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpbQF5Vv143X64cJL:-LLEpbQF5Vv143X64cJM:b-a1efck) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6485)):*
> ```Java
> 
>     private <C extends Column<?, T>> void addColumnsSkipSelectionColumnCheck(Collection<C> columnCollection, int index) {
> //        final Map<Integer, Double> columnWidths = new HashMap<>();
> ```

Dead code to be removed
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6487 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpdwiBLREVYxQ3PqE:-LLEpdwiBLREVYxQ3PqF:b-a1efck) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6487)):*
> ```Java
> //        final Map<Integer, Double> columnWidths = new HashMap<>();
>         int visibleNewColumns = 0;
> //        int currentColumnIndex = index;
> ```

Dead code to be removed
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6500 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEperY8HR0N4Cu4Nt1:-LLEperY8HR0N4Cu4Nt2:b-a1efck) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6500)):*
> ```Java
>             if (!column.isHidden()) {
>                 visibleNewColumns++;
> //                if (column.getWidth() > 0) {
> ```

Dead code to be removed
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6504 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpfo_A7teJXVlHI2I:-LLEpfo_A7teJXVlHI2J:b-a1efck) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6504)):*
> ```Java
> //                }
>             }
> //            currentColumnIndex++;
> ```

Dead code to be removed
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6509 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpgbJ1w06rNvAjOmx:-LLEpgbJ1w06rNvAjOmy:b-a1efck) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6509)):*
> ```Java
>             final ColumnConfiguration columnConfiguration = this.escalator.getColumnConfiguration();
>             columnConfiguration.insertColumns(index, visibleNewColumns);
> //            columnConfiguration.setColumnWidths(columnWidths);
> ```

Dead code to be removed


<!-- Sent from Reviewable.io -->

elmot reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @ericflock)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/11043#-LLEr1UzCR2mc12jioCd:-LLEr1UzCR2mc12jioCe:b-ib6b3d) (no related file):*
The change should be verified using all the visual tests

---


<!-- Sent from Reviewable.io -->

ericflock reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @elmot and @ericflock)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/11043#-LLEr1UzCR2mc12jioCd:-LL_2oiv3TtMmD56Q59V:b-b2lap5) (no related file):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

The change should be verified using all the visual tests
</blockquote></details>

The final change passes the validation build including all framework UI-Tests. Are there any other visual tests I have to run this change on?

---
___
*[client/src/main/java/com/vaadin/client/connectors/GridConnector.java, line 175 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpp9s6WV6y9ObOmY_:-LL_4vyNBlx1ayJxPFkS:biol5lq) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/connectors/GridConnector.java#L175)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

API breaking change. Is it possible to keep old constructor parameters and then get the state via `AbstractRendererConnector.getState()`
</blockquote></details>

I will provide a new commit, for this.
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6485 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpbQF5Vv143X64cJL:-LL_4nus2satTAINw5Fd:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6485)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Dead code to be removed
</blockquote></details>

Done.
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6487 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpdwiBLREVYxQ3PqE:-LL_4oEOCXGl4r4csug7:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6487)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Dead code to be removed
</blockquote></details>

Done.
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6500 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEperY8HR0N4Cu4Nt1:-LL_4okg37VsLSGl49dj:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6500)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Dead code to be removed
</blockquote></details>

Done.
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6504 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpfo_A7teJXVlHI2I:-LL_4pCd5bMwSIdkxnLr:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6504)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Dead code to be removed
</blockquote></details>

Done.
___
*[client/src/main/java/com/vaadin/client/widgets/Grid.java, line 6509 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpgbJ1w06rNvAjOmx:-LL_4peX1LJdH8gGlrbK:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/widgets/Grid.java#L6509)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Dead code to be removed
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ericflock reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @elmot and @ericflock)

___
*[client/src/main/java/com/vaadin/client/connectors/GridConnector.java, line 175 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpp9s6WV6y9ObOmY_:-LLczt5x2LWAUe09Ko93:b-m8n958) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/connectors/GridConnector.java#L175)):*
<details><summary><i>Previously, ericflock wrote…</i></summary><blockquote>

I will provide a new commit, for this.
</blockquote></details>

New commit isn't needed. The change does not break the api.
The CustomGridColumn is an private inner class of the GridConnector. So only the GridConnector can call this constructor or extend the class.

Your suggestion to get the state via AbstractRendererConnector.getState() doesn't work here, because the relevant state is of type GridColumnState.


<!-- Sent from Reviewable.io -->

ericflock reviewed (commented):
> Reviewed 2 of 3 files at r1, 1 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: all files reviewed, 7 unresolved discussions (waiting on @elmot)



<!-- Sent from Reviewable.io -->

ericflock reviewed (commented):
> Reviewed 1 of 3 files at r1, 1 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: all files reviewed, 7 unresolved discussions (waiting on @elmot)



<!-- Sent from Reviewable.io -->

ericflock reviewed (commented):
> Reviewed 1 of 3 files at r1, 1 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: all files reviewed, 7 unresolved discussions (waiting on @elmot)



<!-- Sent from Reviewable.io -->

elmot reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @elmot)

___
*[client/src/main/java/com/vaadin/client/connectors/GridConnector.java, line 175 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpp9s6WV6y9ObOmY_:-LLdIiJ66tiy8o7oK16t:bgso14d) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/connectors/GridConnector.java#L175)):*
<details><summary><i>Previously, ericflock wrote…</i></summary><blockquote>

New commit isn't needed. The change does not break the api.
The CustomGridColumn is an private inner class of the GridConnector. So only the GridConnector can call this constructor or extend the class.

Your suggestion to get the state via AbstractRendererConnector.getState() doesn't work here, because the relevant state is of type GridColumnState.

</blockquote></details>

The public constructor parameters are changed => breaking change in API. Probably we can survive with this, since it is on client side


<!-- Sent from Reviewable.io -->

elmot reviewed (changes requested):
> Reviewed 1 of 3 files at r1, 1 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: all files reviewed, 1 unresolved discussion (waiting on @elmot)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/11043#-LLEr1UzCR2mc12jioCd:-LLdJxW26uwexZ1FUBTj:bn7p97i) (no related file):*
<details><summary><i>Previously, ericflock wrote…</i></summary><blockquote>

The final change passes the validation build including all framework UI-Tests. Are there any other visual tests I have to run this change on?
</blockquote></details>

There is a "full" build process, with a huge pile of tests. It should be run with this change, but  we need to stabilize  the mainstream build first. I am blocking the PR until it's  done and we can move forward.

---


<!-- Sent from Reviewable.io -->

ericflock reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: all files reviewed, 1 unresolved discussion (waiting on @ericflock)

___
*[client/src/main/java/com/vaadin/client/connectors/GridConnector.java, line 175 at r1](https://reviewable.io/reviews/vaadin/framework/11043#-LLEpp9s6WV6y9ObOmY_:-LLdK3R03VHlAZ8oMMBH:bilu80k) ([raw file](https://github.com/vaadin/framework/blob/c40633ac91f3299156eb9ce09f9754651510af19/client/src/main/java/com/vaadin/client/connectors/GridConnector.java#L175)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

The public constructor parameters are changed => breaking change in API. Probably we can survive with this, since it is on client side
</blockquote></details>

Ok... very strict, but I will revert and change it.


<!-- Sent from Reviewable.io -->

ericflock committed:
Make the Grid fast again - avoiding api break

ericflock reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @elmot)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/11043#-LLEr1UzCR2mc12jioCd:-LLdPT7kEAg0dT26hKp8:b-eivrw7) (no related file):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

There is a "full" build process, with a huge pile of tests. It should be run with this change, but  we need to stabilize  the mainstream build first. I am blocking the PR until it's  done and we can move forward.
</blockquote></details>

Ok. I do not exactly know what you mean, because this is my first pull request to vaadin and I do not know how the process works. Please let me know, if or what I need to do.

---


<!-- Sent from Reviewable.io -->

elmot commented:
@ericflock, thanks for your patience, the ball is on our side ATM. People outside the company can not run those validations because of security risks.

ericflock reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/11043#-:-LOqZaxS9D0EE0qYnJEB:bnfp4nl)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @elmot)



<!-- Sent from Reviewable.io -->

TatuLund commented:
Hi @ericflock , we are not going to merge this pull request as is. But it is certainly trying to tackle real issues and work has been already beneficial and motivated us to study this further. However, it would be easier to process this in smaller pieces.

We merged this other patch recently, and it is doing some performance improvement that is proposed also in your patch in less evasive way https://github.com/vaadin/framework/pull/11199 So thus rebasing and refactoring is needed.

There is a way to improve the performance in your patch avoiding unnecessary width calculation of the multiselection checkbox column in your patch. This is something we are still interested in. So if you would be interested to refactor that to separate pull request, that would be most certainly welcome.
